src/kronecker.jl

```diff
@@ -0,0 +1,259 @@
+struct KroneckerMap{T, As<:Tuple{Vararg{LinearMap}}} <: LinearMap{T}
+    maps::As
+    function KroneckerMap{T, As}(maps::As) where {T, As}
+        for A in maps
+            promote_type(T, eltype(A)) == T || throw(InexactError())
+        end
+        return new{T,As}(maps)
+    end
+end
+
+KroneckerMap{T}(maps::As) where {T, As<:Tuple{Vararg{LinearMap}}} = KroneckerMap{T, As}(maps)
+
+"""
+    kron(A::LinearMap, B::LinearMap)
+
+Construct a `KroneckerMap <: LinearMap` object, a (lazy) representation of the
+Kronecker product of two `LinearMap`s. One of the two factors can be an `AbstractMatrix`
+object, which then is promoted to `LinearMap` automatically. To avoid fallback to
+the generic [`Base.kron`](@ref), there must be a `LinearMap` object among the
+first 8 arguments in usage like `kron(A, B, Cs...)`.
+
+Note: If `A`, `B`, `C` and `D` are linear maps of such size that one can form the
+matrix products `A*C` and `B*D`, then the mixed-product property `(A⊗B)*(C⊗D)`
+holds. Upon multiplication of Kronecker products, this rule is checked for
+applicability, which leads to type-instability with a union of two types.
```

Comment:
I should probably read the discussion in this thread, but could you give a quick summary as to why this advantageous over just using `CompositeMap` for the product?

Reply:
Ah, that is a left-over from earlier approaches. Applying the mixed-product rule is advantageous because every Kronecker product allocates a `zero` vector on multiplication, so when you can replace a product of two (or more) Kronecker products by one Kronecker product of products, this saves allocation effort, and maybe further reduces the overall complexity. But the product rule is now implemented in `mul!`, and therefore type stable! I'll remove that sentence from the docstring.